### ISSUE TITLE
[grouping]  Buggy edge cases in `filter_exceptions_for_exception_groups` - no/empty exception `0`

### ISSUE THREAD

**Issue by lobsterkatie** [getsentry/sentry#73592]
TL;DR: In our `chained-exception` strategy, we're not handling the case of one or more outer wrapper errors being gone very well.

Relates to [this `KeyError` event](https://sentry.sentry.io/issues/5552158228/events/9bac15a387104761b1f9066720c48010/), which is erroring out [here](https://github.com/getsentry/sentry/blob/eacf6919c358e584a346605281b0a382de1ecd3a/src/sentry/grouping/strategies/newstyle.py#L802) because there is no entry `0` in `exception_tree` (which is a dict). Values are added to `exception_tree` [a little higher up in the function](https://github.com/getsentry/sentry/blob/eacf6919c358e584a346605281b0a382de1ecd3a/src/sentry/grouping/strategies/newstyle.py#L759-L766), and they're based on `mechanism.exception_id` and `mechanism.parent_id`.

I went and checked out the issue that was being processed, and indeed, neither of the exceptions would put anything at `0`:

![image](https://github.com/getsentry/sentry/assets/14812505/e7246127-5306-4cd5-80df-daa510df26f0)

Seems like we peeled off more than one outer/wrapper exception.

Not high priority, because it's only happened once that we know of, but something we might want to make more robust regardless.

UPDATE: The same project (but a different event, from a different SDK) caused [a related error](https://sentry.sentry.io/issues/5552160394/events/77963f3a03d5404bbd4f30d3976ae2a0/) a few lines later. The JSON was equivalent to the above, except one of the `parent_id`s _was_ `0` (though there was no actual exception `0` in the data). As a result, `exception_tree[0]` was an empty `ExceptionTreeNode` (from [here](https://github.com/getsentry/sentry/blob/eacf6919c358e584a346605281b0a382de1ecd3a/src/sentry/grouping/strategies/newstyle.py#L765)), which then made the [if block creating `top_level_exceptions`](https://github.com/getsentry/sentry/blob/eacf6919c358e584a346605281b0a382de1ecd3a/src/sentry/grouping/strategies/newstyle.py#L802-L807) not run, which then made iterating through it [a few lines later](https://github.com/getsentry/sentry/blob/eacf6919c358e584a346605281b0a382de1ecd3a/src/sentry/grouping/strategies/newstyle.py#L812-L816) error out.
